contracts/TwabRewards.sol

```diff
@@ -287,14 +286,36 @@ contract TwabRewards is ITwabRewards {
         return _promotion;
     }
 
+    /**
+        @notice Compute promotion end timestamp.
+        @param _promotion Promotion to compute end timestamp for
+        @return Promotion end timestamp
+    */
+    function _getPromotionEndTimestamp(Promotion memory _promotion)
+        internal
+        pure
+        returns (uint256)
+    {
+        return _promotion.startTimestamp + (_promotion.epochDuration * _promotion.numberOfEpochs);
+    }
+
     /**
         @notice Get the current epoch id of a promotion.
         @dev Epoch ids and their boolean values are tightly packed and stored in a uint256, so epoch id starts at 0.
-        @dev We calculate the current epoch id if the promotion has started. Otherwise, we return 0.
+        @dev We return the current epoch id if the promotion has not ended.
+        Otherwise, we return the last epoch id or 0 if the promotion has not started.
         @param _promotion Promotion to get current epoch for
         @return Epoch id
      */
     function _getCurrentEpochId(Promotion memory _promotion) internal view returns (uint256) {
+        if (block.timestamp > _getPromotionEndTimestamp(_promotion)) {
```

Comment:
This implemented can be simplified significantly.

Like so:

```
uint256 currentEpochId = (block.timestamp - _promotion.startTimestamp) / epochDuration;
if (currentEpochId >= _promotion.numberOfEpochs) {
  currentEpochId = _promotion.numberOfEpochs - 1;
}
```

HOWEVER - both the current implementation and the above have strange behaviour.  If the current time is beyond the *duration of the last epoch*, then is it still the last epoch?  No.

In this case it's strange that we are still returning the last epoch, even if time has moved beyond the epoch.

In this case, the function needs to do something differently.  Either:

1. Ignore the promotion end.  The end has to be handled by the callers of this function
2. Revert if the current time is greater than the promotion end

Thoughts?  My preference is #1

Reply:
> If the current time is beyond the duration of the last epoch, then is it still the last epoch? No.

It is the last epoch for which we have distributed rewards. At least in the current implementation. 
Which makes sense to me, it means that the rewards stopped at this epoch.

I've tried to implement your suggested simplification but if we do so, a lot of edge cases, like the start of the promotion, are not handled.

Reply:
No, I want the function `_getCurrentEpochId` simplified.

In both use cases you have a guard above that prevents the code from being executed when the promotion has ended.  We can rely on that.

Simplify that function down to one line.  Nested booleans are bad.

Reply:
`uint256 _currentEpochId = (block.timestamp - _promotion.startTimestamp) / _promotion.epochDuration;` will underflow if `block.timestamp` is inferior to `_promotion.startTimestamp`.
It's the case in this test: https://github.com/pooltogether/v4-periphery/blob/1122cd0f3303f8a6744266f4f6df95456c3f458d/test/TwabRewards.test.ts#L250
So we need `if (block.timestamp > _promotion.startTimestamp)`.

If `_currentEpochId` is not superior to `0` it will underflow on `_currentEpochId = _promotion.numberOfEpochs - uint8(1);`.
It's the case in this test: https://github.com/pooltogether/v4-periphery/blob/1122cd0f3303f8a6744266f4f6df95456c3f458d/test/TwabRewards.test.ts#L211
So we need `if (_currentEpochId > 0)`.

I don't see how I can simplify this function more.
https://github.com/pooltogether/v4-periphery/pull/43/commits/edf91582acaf21b3479cbcd0906d73c87adb4271